### PR TITLE
Add getSession and redirectIfLoggedOut

### LINKED ISSUE #9: Add redirectIfLoggedOut method
Add a method to check for the JWT, verify it, and if not present or valid, redirect to a URL.

```js
// Explicitly decide route
Userfront.redirectIfLoggedOut("/login");

// Look up After-logout route
Userfront.redirectIfLoggedOut();
```

### LINKED ISSUE #114: Add a typescript declaration for `Userfront.logout()`
https://github.com/userfront/userfront-core/blob/859a08676e2860cddde8a020fb70368cd4fca392/ts/index.d.ts#L201

There should be a second `logout` definition for when there's no parameters. Right now `Userfront.logout()` triggers an error (you need `Userfront.logout({})`.

### LINKED ISSUE #115: Add isLoggedIn method
We want a method that will allow easy determination of whether a user is logged in.

- [x] Add `isLoggedIn` method
- [x] Update [redirectIfLoggedIn](https://userfront.com/docs/js.html#redirectifloggedin) to use `isLoggedIn`
- [x] Add [redirectIfLoggedOut](https://github.com/userfront/userfront-core/issues/9) using `isLoggedIn`

### PR DESCRIPTION
**Pull request by tyrw**
Normal
Closes #115 
Closes #114 
Closes #9 

- Add the `isLoggedIn` method, which returns `true` if a user is logged in and `false` otherwise
  - If the access token is invalid/expired, will attempt to refresh before returning `false`
- Update `redirectIfLoggedIn` to use `isLoggedIn`
- Add `redirectIfLoggedOut` method (#9)
- Fix typescript declaration for `logout()` (#114)

### CODE REVIEW

File: src/tokens.js [hunk 1]
```diff
@@ -59,6 +60,52 @@ export function unsetTokens() {
...
+  return isJwtLocallyValid(store.tokens.refreshToken);
+}
+
+/**
+ * Determine whether a user is logged in by checking their
+ * JWT access token and, if invalid, refreshing it and checking
+ * again.
+ * @returns {Boolean}
+ */
+export async function isLoggedIn() {
```
Comment (RJFelix) [inline src/tokens.js:87]: Hmm, my instinct is that `isFoo` is a read operation without side effects that returns a plain Boolean. I'd be surprised to call `isLoggedIn` and find that it makes a network request and returns a Promise.

But I haven't thought of a better name that keeps the interface simple... `checkIfLoggedIn` is the best I've got so far.

File: src/tokens.js [hunk 2]
```diff
@@ -59,6 +60,52 @@ export function unsetTokens() {
...
+ */
+export function isRefreshTokenLocallyValid() {
+  return isJwtLocallyValid(store.tokens.refreshToken);
+}
+
+/**
+ * Determine whether a user is logged in by checking their
+ * JWT access token and, if invalid, refreshing it and checking
+ * again.
+ * @returns {Boolean}
```
Comment (RJFelix) [inline src/tokens.js:85]:
```suggestion
 * @returns {Promise<Boolean>}
```

(if this remains async)

File: src/url.js [hunk 3]
```diff
@@ -64,6 +65,31 @@ export async function redirectIfLoggedIn({ redirect } = {}) {
...
+ */
+export async function redirectIfLoggedOut({ redirect } = {}) {
+  // If the user is logged in, return without doing anything
+  const isUserLoggedIn = await isLoggedIn();
+  if (isUserLoggedIn) {
+    return;
+  }
+
+  // Remove all cookies
+  removeAllCookies();
```
Comment (RJFelix) [inline src/url.js:80]: Should the function name & docstring mention this? I think it's fine to remove cookies here, but a client could rely on cookies in a way that we don't expect. Worst case for the client would be a race condition causing flaky behavior if they do `Promise.all([doSomethingWithCookies, redirectIfLoggedOut])`

Like above though, having trouble finding the right balance between simplicity and description in naming. `cleanupAndRedirectIfLoggedOut` is a bit wordy... it's probably OK named as is, if the docstring & documentation calls out that this clears cookies too.

File: src/tokens.js (hunk 1 above)
Comment (moroshko) [reply src/tokens.js:87]: `getUserStatus` could be another idea, could return e.g. `{ isLoggedIn: true, user: { ... } }`.

Comment (tyrw) [reply src/tokens.js:87]: I agree with the sentiment. It started as a boolean and then evolved into a promise, so the name should probably change. I'll put something up shortly.

File: src/tokens.js (hunk 2 above)
Comment (tyrw) [reply src/tokens.js]: Updated, thanks!

File: src/url.js (hunk 3 above)
Comment (tyrw) [reply src/url.js:80]: Despite the name, this only removes cookies related to the session: access, refresh, and id. If we have determined that the user is no longer logged in, then those cookies should not be considered valid.

- [ ] Will make sure to note that it cleans up the cookies in this case in the documentation.

File: src/tokens.js (hunk 1 above)
Comment (tyrw) [reply src/tokens.js]: Ended up going with `getSession`, which can evolve to give lots of information about the session (duration, original location, number of times refreshed, factor(s) used for login, etc.). For now, usage looks like:

```
const { isLoggedIn } = await Userfront.getSession();
```

@moroshko I published this as `@userfront/core@0.4.3-beta.0` pending your feedback.

File: src/utils.js [hunk 4]
```diff
@@ -22,6 +22,28 @@ export function getJWTPayload(token) {
...
+export function isJwtLocallyValid(token) {
+  try {
+    // Must be present
+    if (!token) {
+      return false;
+    }
+
+    // Must not be expired
+    const payload = getJwtPayload(token);
+    return new Date(payload.exp) > new Date();
```
Comment (moroshko) [inline src/utils.js:41]: @tyrw I believe `payload.exp` should be `payload.exp * 1000`. 
This function returns `false` in its current form even when the jtw hasn't expired.

File: test/config/utils.js [hunk 5]
```diff
@@ -36,7 +36,7 @@ const sharedTokenProperties = {
   token: {
     sessionId: "7b794d68-dc94-4829-bd68-596c5d2693b5",
     iat: new Date().getTime(),
-    exp: new Date().getTime(),
+    exp: new Date().getTime() + 86400000, // 1 day
```
Comment (moroshko) [inline test/config/utils.js:39]: @tyrw These should be in seconds, not milliseconds. See `NumericDate` definition [here](https://www.rfc-editor.org/rfc/rfc7519#page-6).

Comment (tyrw) [reply test/config/utils.js]: Thanks!

File: src/utils.js (hunk 4 above)
Comment (tyrw) [reply src/utils.js]: This is correct, and fixed with 0.4.6-beta.0

**Comment by tyrw** [comment]
Published as 0.4.6

**Review by RJFelix — APPROVED** [review]
